doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+# Metadata database for Materialize Platform
+
+## Summary
+
+Each [Materialize Platform layer](../platform/architecture-db.md) in each
+customer environment in Materialize Platform needs to durably record its
+metadata. What database to use is an open question. The correctness,
+performance, scalability, reliability, and availability goals of Materialize
+Platform heavily constrain our options. This document details those constraints
+and proposes a candidate system for further evaluation.
+
+## Constraints
+
+### Hard requirements
+
+The following are the non-negotiable requirements that flow from Materialize
+Platform's product goals.
+
+* **Resilience to single availability zone outages.** Materialize Platform must
+  be resilient to an outage in a single availability zone of a cloud provider
+  region. Therefore the metadata database must be resilient to outages in a
+  single availability zone.
+
+* **Strong consistency.** Materialize Platform intends to offer strict
+  serializability, at least in some configurations, that withstands Jepsen
+  analysis. This imposes the same constraints on the underlying metadata
+  database.
+
+* **Single-millisecond commit latency.** The metadata database will be on the
+  critical path for all write queries issued against Materialize Platform, and
+  (likely) all read queries in strictly serializable mode. Transactions against
```

Comment:
Maybe this doesn't change much in our decision but I thought it was worth a mention.

Under normal operation the ADAPTER layer will only be performing write operations and never read operations, because it is the only writer and can hold the state in-memory (I think this is true with sqlite today, right?). The only situation when a strictly serializable read must happen is during recovery after a restart.

If the above are true then a system that does as little as possible on write (e.g just appending to a log) and a lot of work on read (e.g have to replay the log) could be simpler to operate while offering the required performance. A lazily compacted kafka queue for example could serve as a backing store for a `Stash` collection.

Reply:
> The above is only true if you can guarantee that it's the only writer instance

Yeah, that's why I mentioned above that the storage layer needs to support fencing. For example kafka is such a system that both supports fencing and is write-optimized.

>  Please be aware that Kubernetes does not do any fencing

Well, kubernetes does have fencing on `StatefulSets`, there can never be more than one running at a time. I don't expect us to use them due to their bad characteristics in case of failures but they definitely guarantee that there will at most one copy of your pod running at all times.

Reply:
> Well, kubernetes does have fencing on StatefulSets, there can never be more than one running at a time

I thought part of the problem could be things like a delayed packet (Kleppmann discusses github's 90 second packet delay in his blog post about locks), so even things like a statefulset don't protect against this.

Reply:
> I thought part of the problem could be things like a delayed packet

This is a problem that even a single threaded program that is running on a single machine needs to take care of. Like, imagine the single thread sending a message, crashing, and restarting (on the same machine). It still needs to deal with the fact that there is potentially a network message in transit even though there was no concurrency. For example it could have written to its local disk that it is about to send a network request before sending it or something.

k8s `StatefulSet`s provide you with that programming model (where thread is replaced with pod), but at most a single machine at a time nonetheless. Additionally it guarantees that whatever persistent state the previous process wrote to its local disk will be visible to the next process that gets spawned up. So techniques like using a WAL can be used here too.

So in the end the issues you have to deal with in a `StatefulSet` setup are exactly the same as a single-threaded program running on a single machine where it can crash and restart.

Reply:
To try to put it succinctly: you need anything that's storing state to participate in fencing. If you're using StatefulSets with attached volumes, Kubernetes does that fencing for you. If you're using anything else (e.g., a networked database), you're on the hook for providing fencing.

Since we want to move away from attached volumes due to their operational complexity, we're switching to a networked database, and now we need to provide teh fencing.

> * **Single-millisecond commit latency.** The metadata database will be on the critical path for all write queries issued against Materialize Platform, and (likely) all read queries in strictly serializable mode.

Sorry, @petrosagg, I think I confused you with the phrasing here. What I meant is that if you're running _Materialize_ in strictly serializable mode—which is not a thing that exists yet—we may need to write to the metadata store to e.g. record the new highest timestamp we've ever sent to a client. Not that we needed to perform a strictly serializable read against the metadata store.

Reply:
I think we're equivocating on the term `fencing`. Fencing is the act of guaranteeing that the old thing cannot interact with the system anymore. It relates to killing or network cordoning old things, not the spawning of new ones.

For clarity, Kubernetes does not do fencing (guarantee the old one is gone), even with statefulsets. It does nothing to ensure the previously existing (possible unknown state) pod is dead. What it does do, is refuse to start another pod to replace it if it can't guarantee that the old one is gone. This gives us the safety property we need, at the expense of availability. If it can't guarantee the pod is gone (ie: in the case of a node failure), we will just be stuck with that pod being down.

An example of fencing would be terminating the AWS EC2 instance running the pod. There are Kubernetes operators that can do this for us if we need, but K8S does not do this kind of thing by default. Even with that kind of fencing, there is still possibility of things like massively delayed packets.

Example situation:
1. Kubernetes loses contact with a node.
2. Pods running on that node go into unknown state.
3. Non-statefulset pods get rescheduled after some delay.
4. Statefulset pods stay in unknown state until a human intervenes.